src/btree/bt_vrfy.c

```diff
@@ -594,84 +588,94 @@ __verify_addr_ts(WT_SESSION_IMPL *session, WT_REF *ref, WT_CELL_UNPACK *unpack,
+    prev_btree_id = 0;      /* [-Wconditional-uninitialized] */
+    uri_itr = NULL;
 
     WT_ERR(__wt_scr_alloc(session, 0, &tmp));
     WT_ERR(__wt_hs_cursor(session, &session_flags));
     cursor = session->hs_cursor;
 
-    while ((ret = cursor->next(cursor)) == 0) {
+    /*
+     * If a uri has been provided, limit verification to the corresponding btree by jumping to the
+     * first record for that btree in the history store. Otherwise scan the whole history store.
+     */
+    if (uri != NULL) {
+        ret = __wt_metadata_uri_to_btree_id(session, uri, &btree_id_given_uri);
+        if (ret != 0)
+            WT_ERR_MSG(session, ret, "Unable to locate the URI %s in the metadata file.", uri);
+
+        /*
+         * Position the cursor at the first record of the specified btree, or one after. It is
+         * possible there are no records in the history store for this btree.
+         */
+        cursor->set_key(cursor, btree_id_given_uri, &hs_key, 0, 0, 0, 0);
+        ret = cursor->search_near(cursor, &exact);
+        if (ret == 0 && exact < 0)
+            ret = cursor->next(cursor);
+        if (ret == WT_NOTFOUND) {
+            ret = 0;
+            goto err;
+        }
+        WT_ERR(ret);
+    } else
+        ret = cursor->next(cursor);
+
+    /* We have the history store cursor positioned at the first record that we want to verify. */
+    for (; ret == 0; ret = cursor->next(cursor)) {
         WT_ERR(cursor->get_key(cursor, &btree_id, &hs_key, &hs_start.timestamp, &hs_start.txnid,
           &hs_stop.timestamp, &hs_stop.txnid));
+
+        /* When limiting our verification to a uri, bail out if the btree-id doesn't match. */
+        if (uri != NULL && btree_id != btree_id_given_uri)
+            break;
+
         /*
          *  Keep track of the previous comparison. The history store is stored in order, so we can
          *  avoid redundant comparisons. Previous btree ID isn't set, until data cursor is open.
          */
         if (data_cursor == NULL || (prev_btree_id != btree_id)) {
             /*
-             * Find the URI from the metadata and validate the btree ID. Using this URI, verify the
-             * history store key with the data store.
+             * Check whether this btree-id exists in the metadata. We do that by finding what uri
+             * this btree belongs to. Using this URI, verify the history store key with the data
+             * store.
              */
             if (data_cursor != NULL) {
                 WT_ERR(data_cursor->close(data_cursor));
                 /* Setting data_cursor to null, to avoid double free */
                 data_cursor = NULL;
             }
-            WT_ERR(__verify_btree_id_with_meta(session, btree_id, &uri));
-            WT_ERR(__wt_open_cursor(session, uri, NULL, NULL, &data_cursor));
+            __wt_free(session, uri_itr);
+            ret = __wt_metadata_btree_id_to_uri(session, btree_id, &uri_itr);
```

Comment:
What is `uri_itr` short for?

Reply:
I added a comment to explain what that function is doing. `uri_itr` is for uri_iterator, for each entry in the history store, if it is for a btree we have not seen before, we check if there is an entry for that data store in the metadata and get the uri for it. The uri then gets used to report the error if one if one is found eventually.